### PR TITLE
#57109 Verification of connection to apiserver in kube-proxy

### LINKED ISSUE #57109: kube-proxy should not be "ready" if it can't reach the apiserver
/kind feature

**What happened**:
When for whatever reason kube-proxy can't reach the apiserver, it retries but the pod stays in ready state.

**What you expected to happen**:
kube-proxy should fail or have a health check that would indicate such issues.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8.4

### PR DESCRIPTION
**Pull request by antoniaklja**
**What this PR does / why we need it**:
Extended kube-proxy healthcheck to verify connection with apiserver.

regarding what @discordianfish said:
> When for whatever reason kube-proxy can't reach the apiserver, it retries but the pod stays in ready state.

> kube-proxy should fail or have a health check that would indicate such issues.

**Which issue(s) this PR fixes** 
Fixes #57109

### CODE REVIEW

**Comment by antoniaklja** [comment]
Also I would like to add some unit and e2e tests in next few days.

**Comment by antoniaklja** [comment]
hey @discordianfish I've added unit test with fakeClient but not sure how we can test failure flow

**Comment by antoniaklja** [comment]
@lavalamp @dchen1107 @discordianfish  ready to review

**Comment by jhorwit2** [comment]
How is this going to impact services that rely on the kube-proxy healthcheck like cloud load balancers? I would not expect or want to have kube-proxy start failing healthchecks which causes a complete outage on all our load balancer services that don't use `externalTrafficPolicy: Local`.

File: pkg/proxy/healthcheck/healthcheck_test.go [hunk 1]
```diff
@@ -385,6 +388,10 @@ func TestHealthzServer(t *testing.T) {
 	testHealthzHandler(server, http.StatusOK, t)
 }
 
+func newFakeClient() clientset.Interface {
```
Comment (jhorwit2) [inline pkg/proxy/healthcheck/healthcheck_test.go]: This check will cause outages for cloud provider load balancer services that already have or are being told to move to the kube-proxy healthz port for default healthchecks as discussed [here](https://github.com/kubernetes/kubernetes/issues/14661#issuecomment-348332432) during apiserver outages.

**Review by jhorwit2 — CHANGES_REQUESTED** [review]

**Comment by jhorwit2** [comment]
cc @MrHohn

**Comment by antoniaklja** [comment]
thanks @jhorwit2, I haven't known that. Looks like side effect of it is quite dangerous and we should figure out something different or close this PR.

**Comment by jhorwit2** [comment]
Agreed @antoniaklja. I think if this is something we need for kube-proxy then we may want to separate the health checks. Perhaps keep 10256 for cloud health and have a new port for kube proxy health flso we keep backwards compatibility.

**Comment by antoniaklja** [comment]
yeah, separated health check just for cloud load balancers sounds fine because cluster in general it'll be more fault tolerant but on the other hand you know if something is going on with connection between kube-roxy and apiserver.

**Comment by discordianfish** [comment]
@jhorwit2 I'm probably missing something here but why would a cloud load balancer be configured to check the kube-proxy health endpoint? I though kube-proxy is used only within the cluster and the cloud load balancer would access ports mapped by kube-proxy and used health endpoints exposed by kube-proxy but not kube-proxy's health endpoint itself?

Either way, my main concern is that kube-proxy starts up and resports as health even if it *never* reached the apiserver. At least in this case it should be save for the health endpoint to return failure, since kube-proxy can't do anything without getting some state first from apiserver.

**Comment by jhorwit2** [comment]
@discordianfish The original discussion around all that is here https://github.com/kubernetes/kubernetes/issues/14661. 

I understand where you are coming from but still have concerns especially when kube-proxy is running as a daemonset. This means that during an apiserver outage it will start failing kubelet checks for liveness and readiness.

**Comment by discordianfish** [comment]
@jhorwit2 Oh I share your concern. Just as of today there doesn't seem to be anything depending on the health check. I agree that this would be a problem if #14661 uses kube-proxy's health endpoint.
So I don't care much about the implementation details, just that it's not reporting "health" if it never reached the apiserver.

**Comment by antoniaklja** [comment]
what is the outcome of this discussion? It looks like guys from #14661 are implementing separate health checks for every cloud provider.

**Comment by dims** [comment]
@antoniaklja @discordianfish can we please close this? PR was last rebased a while ago.

**Pull request closed without merging**